code/krepel/math/vector3.d

```diff
@@ -0,0 +1,569 @@
+module krepel.math.vector3;
+
+import krepel.math.math;
+import krepel.algorithm.comparison;
+import krepel.math.vector2;
+import krepel.math.vector4;
+import std.conv;
+
+@nogc:
+@safe:
+nothrow:
+
+/// Calculates the Dot Product of the two Vectors
+/// Input Vectors will not be modified
+float Dot(Vector3 Lhs, Vector3 Rhs)
+{
+  float Result = Lhs | Rhs;
+  return Result;
+}
+
+/// Multiplies two Vectors per component returning a new Vector containing
+/// (Lhs.X * Rhs.X, Lhs.Y * Rhs.Y, Lhs.Z * Rhs.Z)
+/// Input Vectors will not be modified
+Vector3 Mul(Vector3 Lhs, Vector3 Rhs)
+{
+  return Lhs * Rhs;
+}
+
+/// Calculates the cross Product (Lhs x Rhs) and returns the result
+/// Input Vectors will not be modified
+Vector3 Cross(Vector3 Lhs, Vector3 Rhs)
+{
+  return Lhs ^ Rhs;
+}
+
+/// Calculates the squared length  of the given Vector, which is the same as the dot product with the same Vector
+/// Input Vector will not be modified
+float LengthSquared(Vector3 Vec)
+{
+  return Vec | Vec;
+}
+
+/// Calculates the squared length of the X and Y components, ignoring the Z component
+/// Input Vector will not be modified
+float LengthSquared2D(Vector3 Vec)
+{
+  return Vec.X * Vec.X + Vec.Y * Vec.Y;
```

Comment:
Technically yes, but then the compiler cannot optimize this:

Dissasembly not using swizzle (note the SIMD instructions):
```
    42: 
    43: /// Calculates the squared length of the X and Y components, ignoring the Z component
    44: /// Input Vector will not be modified
    45: float LengthSquared2D(Vector3 Vec)
00007FF733460E70  push        rbp  
00007FF733460E71  mov         rbp,rsp  
00007FF733460E74  mov         qword ptr [Vec],rcx  
    46: {
    47:   return Vec.X * Vec.X + Vec.Y * Vec.Y;
00007FF733460E78  mov         rax,qword ptr [Vec]  
00007FF733460E7C  movss       xmm0,dword ptr [rax]  
00007FF733460E80  movss       xmm1,dword ptr [rax]  
00007FF733460E84  mulss       xmm0,xmm1  
00007FF733460E88  lea         rcx,[rax+4]  
00007FF733460E8C  movss       xmm2,dword ptr [rcx]  
00007FF733460E90  movss       xmm3,dword ptr [rcx]  
00007FF733460E94  mulss       xmm2,xmm3  
00007FF733460E98  addss       xmm0,xmm2  
    48: }
00007FF733460E9C  pop         rbp  
00007FF733460E9D  ret  
```

And using swizzle:
```
    42: 
    43: /// Calculates the squared length of the X and Y components, ignoring the Z component
    44: /// Input Vector will not be modified
    45: float LengthSquared2D(Vector3 Vec)
00007FF6FC66C140  push        rbp  
00007FF6FC66C141  mov         rbp,rsp  
00007FF6FC66C144  sub         rsp,10h  
00007FF6FC66C148  mov         qword ptr [Vec],rcx  
    46: {
    47:   return Vec.XY | Vec.XY;
00007FF6FC66C14C  mov         rcx,qword ptr [Vec]  
00007FF6FC66C150  sub         rsp,20h  
00007FF6FC66C154  call        _D6krepel4math7vector37Vector327__T10opDispatchVAyaa2_5859Z10opDispatchMNgFNbNiNfZS6krepel4math7vector27Vector2 (07FF6FC666311h)  
00007FF6FC66C159  add         rsp,20h  
00007FF6FC66C15D  mov         qword ptr [rbp-10h],rax  
00007FF6FC66C161  lea         rax,[rbp-10h]  
00007FF6FC66C165  mov         qword ptr [rbp-8],rax  
00007FF6FC66C169  mov         rcx,qword ptr [Vec]  
00007FF6FC66C16D  sub         rsp,20h  
00007FF6FC66C171  call        _D6krepel4math7vector37Vector327__T10opDispatchVAyaa2_5859Z10opDispatchMNgFNbNiNfZS6krepel4math7vector27Vector2 (07FF6FC666311h)  
00007FF6FC66C176  add         rsp,20h  
00007FF6FC66C17A  mov         rdx,rax  
00007FF6FC66C17D  mov         rcx,qword ptr [rbp-8]  
00007FF6FC66C181  sub         rsp,20h  
00007FF6FC66C185  call        _D6krepel4math7vector27Vector223__T8opBinaryHVAyaa1_7cZ8opBinaryMFNaNbNiNfS6krepel4math7vector27Vector2Zf (07FF6FC6650D3h)  
00007FF6FC66C18A  add         rsp,20h  
    48: }
00007FF6FC66C18E  mov         rsp,rbp  
00007FF6FC66C191  pop         rbp  
00007FF6FC66C192  ret  
```